parquet-hadoop/src/main/java/org/apache/parquet/hadoop/util/CompatibilityUtil.java

```diff
@@ -0,0 +1,396 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.parquet.hadoop.util;
+
+import org.apache.hadoop.fs.FSDataInputStream;
+import java.io.EOFException;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.util.EnumSet;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.lang.reflect.Constructor;
+
+import org.apache.parquet.Log;
+import parquet.org.apache.thrift.TBase;
+import parquet.org.apache.thrift.TException;
+import parquet.org.apache.thrift.protocol.*;
+import parquet.org.apache.thrift.transport.TIOStreamTransport;
+import parquet.org.apache.thrift.transport.TTransport;
+import parquet.org.apache.thrift.transport.TTransportException;
+
+public class CompatibilityUtil {
+  private static final boolean useV21;
+  private static final Log LOG = Log.getLog(CompatibilityUtil.class);
+  public static final V21FileAPI fileAPI;
+  private static final int MAX_SIZE = 1 << 20;
+  private static final Object bufferPool;
+  
+  private static class V21FileAPI {
+    private final Constructor<?> ELASTIC_BYTE_BUFFER_CONSTRUCTOR;
+    private final Class<?> ElasticByteBufferCls;
+    private final Class<?> ByteBufferCls;
+    private final Class<? extends Enum> ReadOptionCls;
+    private final Method READ_METHOD;
+    private final Method RELEASE_BUFFER_METHOD;
+    private final Method GET_BUFFER_METHOD;
+    private final Method PUT_BUFFER_METHOD;
+    private final Class<?> FSDataInputStreamCls;
+    
+    private V21FileAPI() throws ClassNotFoundException, NoSuchMethodException, SecurityException {
+      final String PACKAGE = "org.apache.hadoop";
+      ElasticByteBufferCls = Class.forName(PACKAGE + ".io.ElasticByteBufferPool");
+      ELASTIC_BYTE_BUFFER_CONSTRUCTOR = ElasticByteBufferCls.getConstructor();
+      ByteBufferCls = Class.forName(PACKAGE + ".io.ByteBufferPool");
+      FSDataInputStreamCls = Class.forName(PACKAGE + ".fs.FSDataInputStream");
+      ReadOptionCls = (Class<Enum>)Class.forName(PACKAGE + ".fs.ReadOption");
+      READ_METHOD = FSDataInputStreamCls.getMethod("read", ByteBufferCls, int.class, EnumSet.class);
+      RELEASE_BUFFER_METHOD = FSDataInputStreamCls.getMethod("releaseBuffer", ByteBuffer.class);
+      GET_BUFFER_METHOD = ElasticByteBufferCls.getMethod("getBuffer", boolean.class, int.class);
+      PUT_BUFFER_METHOD = ElasticByteBufferCls.getMethod("putBuffer", ByteBuffer.class);
+    }
+  }
+  
+  static {
+    boolean v21 = true;
+    try {
+      Class.forName("org.apache.hadoop.io.ElasticByteBufferPool");
+    } catch (ClassNotFoundException cnfe) {
+      v21 = false;
+    }
+    
+    useV21 = v21;
```

Comment:
I have remove a lot of the content of this class as the methods were unused. I did end up leaving this flag here to avoid the exception based control flow used to handle situations where the interface was available but only threw unsupported operation exception. Now there is only one method needed in this class and I call it in this static block and appropriately set useV21 to false to avoid calling the method later and instead use the legacy API.